File: killrvideo-service-comments/src/main/java/com/killrvideo/service/comment/grpc/CommentsServiceGrpcMapper.java
package com.killrvideo.service.comment.grpc;

import com.killrvideo.dse.dto.ResultListPage;
import com.killrvideo.service.comment.dto.Comment;
import com.killrvideo.service.comment.dto.QueryCommentByUser;
import com.killrvideo.service.comment.dto.QueryCommentByVideo;
import com.killrvideo.utils.GrpcMappingUtils;
import killrvideo.comments.CommentsServiceOuterClass;
import killrvideo.comments.CommentsServiceOuterClass.*;
import killrvideo.comments.events.CommentsEvents;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.Optional;

import static com.killrvideo.utils.GrpcMappingUtils.*;
import static org.apache.commons.lang3.StringUtils.isBlank;
import static org.apache.commons.lang3.StringUtils.isNotBlank;

/**
 * Validation of inputs and mapping
 *
 * @author DataStax Developer Advocates team.
 */
@Component
public class CommentsServiceGrpcMapper {
    // --- Mappings ---
    
    /**
     * Utility from exposition to Dse query.
     * 
     * @param grpcReq
     *      grpc Request
     * @return
     *      query bean for Dao
     */
    public QueryCommentByUser mapFromGrpcUserCommentToDseQuery(GetUserCommentsRequest grpcReq) {
        QueryCommentByUser targetQuery = new QueryCommentByUser();
        if (grpcReq.hasStartingCommentId() && 
                !isBlank(grpcReq.getStartingCommentId().getValue())) {
            targetQuery.setCommentId(Optional.of(fromTimeUuid(grpcReq.getStartingCommentId())));
        }
        targetQuery.setUserId(fromUuid(grpcReq.getUserId()));
        targetQuery.setPageSize(grpcReq.getPageSize());
        if (isNotBlank(grpcReq.getPagingState())) {
            targetQuery.setPageState(Optional.of(grpcReq.getPagingState()));
        }
        return targetQuery;
    }
    
    // Map from CommentDseDao response bean to expected GRPC object.
    public GetVideoCommentsResponse mapFromDseVideoCommentToGrpcResponse(ResultListPage<Comment> comments) {
        final GetVideoCommentsResponse.Builder builder = GetVideoCommentsResponse.newBuilder();
        for (Comment c : comments.getResults()) {
           builder.setVideoId(uuidToUuid(c.getVideoid()));
           builder.addComments(newVideoCommentProto(c));
        }
        comments.getPagingState().ifPresent(builder::setPagingState);
        return builder.build();
    }

    private CommentsServiceOuterClass.VideoComment newVideoCommentProto(Comment c) {
        return CommentsServiceOuterClass.VideoComment.newBuilder()
                .setComment(c.getComment())
                .setUserId(uuidToUuid(c.getUserid()))
                .setCommentId(uuidToTimeUuid(c.getCommentid()))
                .setCommentTimestamp(instantToTimeStamp(c.getDateOfComment()))
                .build();
    }

    // Map from CommentDseDao response bean to expected GRPC object.
    public GetUserCommentsResponse mapFromDseUserCommentToGrpcResponse(ResultListPage<Comment> dseRes) {
        final GetUserCommentsResponse.Builder builder = GetUserCommentsResponse.newBuilder();
        for (Comment c : dseRes.getResults()) {
           builder.setUserId(uuidToUuid(c.getUserid()));
           builder.addComments(newUserCommentProto(c));
        }
        dseRes.getPagingState().ifPresent(builder::setPagingState);
        return builder.build();
    }

    private CommentsServiceOuterClass.UserComment newUserCommentProto(Comment c) {
        return CommentsServiceOuterClass.UserComment.newBuilder()
                .setComment(c.getComment())
                .setCommentId(uuidToTimeUuid(c.getCommentid()))
                .setVideoId(uuidToUuid(c.getVideoid()))
                .setCommentTimestamp(instantToTimeStamp(c.getDateOfComment()))
                .build();
    }

    /**
     * Utility from exposition to Dse query.
     * 
     * @param grpcReq
     *      grpc Request
     * @return
     *      query bean for Dao
     */
    public QueryCommentByVideo mapFromGrpcVideoCommentToDseQuery(GetVideoCommentsRequest grpcReq) {
        QueryCommentByVideo targetQuery = new QueryCommentByVideo();
        if (grpcReq.hasStartingCommentId() && 
                !isBlank(grpcReq.getStartingCommentId().getValue())) {
            targetQuery.setCommentId(Optional.of(fromTimeUuid(grpcReq.getStartingCommentId())));
        }
        targetQuery.setVideoId(fromUuid(grpcReq.getVideoId()));
        targetQuery.setPageSize(grpcReq.getPageSize());
        if (isNotBlank(grpcReq.getPagingState())) {
            targetQuery.setPageState(Optional.of(grpcReq.getPagingState()));
        }
        return targetQuery;
    }

    public Comment mapToComment(CommentOnVideoRequest grpcReq) {
        Comment comment = new Comment();
        comment.setVideoid(fromUuid(grpcReq.getVideoId()));
        comment.setCommentid(fromTimeUuid(grpcReq.getCommentId()));
        comment.setUserid(fromUuid(grpcReq.getUserId()));
        comment.setComment(grpcReq.getComment());
        return comment;
    }

    public CommentsEvents.UserCommentedOnVideo createUserCommentedOnVideoEvent(Comment comment) {
        return CommentsEvents.UserCommentedOnVideo.newBuilder()
                .setCommentId(uuidToTimeUuid(comment.getCommentid()))
                .setVideoId(uuidToUuid(comment.getVideoid()))
                .setUserId(uuidToUuid(comment.getUserid()))
                .setCommentTimestamp(GrpcMappingUtils.instantToTimeStamp(Instant.now()))
                .build();
    }
}

File: killrvideo-service-comments/src/main/kotlin/com/killrvideo/service/comment/grpc/CommentsServiceGrpcMapper.kt
package com.killrvideo.service.comment.grpc

import com.killrvideo.dse.dto.ResultListPage
import com.killrvideo.service.comment.dto.Comment
import com.killrvideo.service.comment.dto.QueryCommentByUser
import com.killrvideo.service.comment.dto.QueryCommentByVideo
import com.killrvideo.utils.GrpcMappingUtils.*
import killrvideo.comments.CommentsServiceOuterClass.*
import killrvideo.comments.events.CommentsEvents.UserCommentedOnVideo
import killrvideo.comments.events.userCommentedOnVideo
import killrvideo.comments.getUserCommentsResponse
import killrvideo.comments.getVideoCommentsResponse
import killrvideo.comments.userComment
import killrvideo.comments.videoComment
import org.apache.commons.lang3.StringUtils.isBlank
import org.apache.commons.lang3.StringUtils.isNotBlank
import org.springframework.stereotype.Component
import java.time.Instant

/**
 * Validation of inputs and mapping
 *
 * @author DataStax Developer Advocates team.
 */
@Component
class CommentsServiceGrpcMapper {
    // --- Mappings ---
    object GetUserCommentsRequestExtensions {
        fun GetUserCommentsRequest.parse(): QueryCommentByUser =
            QueryCommentByUser(
                userId = fromUuid(this.userId),
                pageSize = this.pageSize,
                pageState = if (isNotBlank(this.pagingState)) this.pagingState else null,
                commentId = if (this.hasStartingCommentId() && !isBlank(this.startingCommentId.value))
                    fromTimeUuid(this.startingCommentId)
                else null
            )
    }

    object GetVideoCommentsRequestExtensions {
        fun GetVideoCommentsRequest.parse(): QueryCommentByVideo =
            QueryCommentByVideo(
                videoId = fromUuid(this.videoId),
                pageSize = this.pageSize,
                pageState = if (isNotBlank(this.pagingState))
                    this.pagingState
                else null,
                commentId = if (this.hasStartingCommentId() && !isBlank(this.startingCommentId.value))
                    fromTimeUuid(this.startingCommentId)
                else null
            )
    }

    object CommentOnVideoRequestExtensions {
        fun CommentOnVideoRequest.parse(): Comment =
            Comment(
                videoid = if (hasVideoId()) fromUuid(this.videoId) else null,
                commentid = if (hasCommentId()) fromTimeUuid(this.commentId) else null,
                userid = if (hasUserId()) fromUuid(this.userId) else null,
                comment = if (hasCommentId()) this.comment else null,
                dateOfComment = Instant.now()
            )
    }

    // Map from CommentDseDao response bean to expected GRPC object.
    fun mapFromDseVideoCommentToGrpcResponse(comments: ResultListPage<Comment>): GetVideoCommentsResponse =
        getVideoCommentsResponse {
            for (c in comments.results) {
                c.videoid?.let { videoId = uuidToUuid(it) }
                this.comments.add(newVideoCommentProto(c))
            }
            comments.pagingState.ifPresent {
                pagingState = it
            }
        }

    private fun newVideoCommentProto(c: Comment): VideoComment =
        videoComment {
            c.comment?.let { comment = it }
            c.userid?.let { userId = uuidToUuid(it) }
            c.commentid?.let { commentId = uuidToTimeUuid(it) }
            c.dateOfComment?.let { commentTimestamp = instantToTimeStamp(it) }
        }

    // Map from CommentDseDao response bean to expected GRPC object.
    fun mapFromDseUserCommentToGrpcResponse(dseRes: ResultListPage<Comment>): GetUserCommentsResponse =
        getUserCommentsResponse {
            for (c in dseRes.results) {
                c.userid?.let { userId = uuidToUuid(it) }
                comments.add(newUserCommentProto(c))
            }
            dseRes.pagingState.ifPresent { pagingState = it }
        }

    private fun newUserCommentProto(c: Comment): UserComment =
        userComment {
            c.comment?.let { comment = it }
            c.commentid?.let { commentId = uuidToTimeUuid(it) }
            c.videoid?.let { videoId = uuidToUuid(it) }
            c.dateOfComment?.let { commentTimestamp = instantToTimeStamp(it) }
        }

    fun createUserCommentedOnVideoEvent(comment: Comment): UserCommentedOnVideo =
        userCommentedOnVideo {
            comment.commentid?.let { commentId = uuidToTimeUuid(it) }
            comment.videoid?.let { videoId = uuidToUuid(it) }
            comment.userid?.let { userId = uuidToUuid(it) }
            comment.dateOfComment?.let { commentTimestamp = instantToTimeStamp(it) }
        }
}
